js/repl/ASTPanel.js

```diff
@@ -0,0 +1,167 @@
+// @flow
+
+import { css } from "emotion";
+import { colors } from "./styles";
+import React from "react";
+import ReactJson from "react-json-view";
+import {
+  flatten,
+  unflatten,
+  filterFlatten,
+  deleteFlatten,
+  mergeFlatten,
+} from "./ASTUtils";
+
+type Props = {
+  className?: string,
+  src: Object,
+};
+
+type State = {
+  src: Object,
+  flattenSrc: Object,
+  flattenType: Object,
+  astOption: {
+    autofocus: boolean,
+    location: boolean,
+    empty: boolean,
+    type: boolean,
+  },
+};
+
+const OPTION_ORDER = ["autofocus", "location", "empty", "type"];
+
+export default class ASTPanel extends React.Component<Props, State> {
+  state = {
+    src: {},
+    flattenSrc: {},
+    flattenType: {},
+    astOption: {
+      autofocus: true,
+      location: true,
+      empty: true,
+      type: true,
+    },
+  };
+
+  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+    const flattenSrc = flatten(nextProps.src);
+    if (nextProps.src !== prevState.src) {
+      return {
+        src: unflatten(flattenSrc),
+        flattenSrc: flattenSrc,
+        flattenType: filterFlatten(flattenSrc, "type"),
+      };
+    }
+    return null;
+  }
+
+  onOptionSettingCheck(option: string) {
+    this.setState(prevState => ({
+      astOption: {
+        ...prevState.astOption,
+        [option]: !prevState.astOption[option],
+      },
+    }));
+
+    this.onChangeJson(option);
```

Comment:
Since `onChangeJson` deals with the updated state of above `setState`, this calls should be in the callback of `setState` above.